ci/test_commit_message.py

```diff
@@ -9,8 +9,18 @@
 import re
 import sys
 
+
 # This script is written to be used with CI Integration
 
+def hasUrl(string):
+    # findall() has been used
+    # with valid conditions for urls in string
+    regex = r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.]" \
+            r"[a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))" \
+            r"+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))"
+    url = re.findall(regex, string)
```

Comment:
When I tested this in the python repl, I got some empty hits with this regex but my URL was detected correctly and the overrall length of url works. Is this expected behavior?

```
>>> msgtxt
'This is a boguz commit message for testing\n\nthis is legit\n\n[4]: https://gist.github.com/JamieMagee/d9cf1ca579b230db504186cbae816f2d#file-distroless-python3-debug-log\n'
>>> regex = r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.]" \
...     r"[a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))" \
...     r"+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))"
>>> 
>>> url = re.findall(regex, msgtxt)
>>> url
[('https://gist.github.com/JamieMagee/d9cf1ca579b230db504186cbae816f2d#file-distroless-python3-debug-log', '', '', '', '')]
>>> len(url)
1

```

Reply:
Even with a simple URL like `https://www.google.com` I see the four additional blank strings in the tuple in the `url` list.

Reply:
Yes, this is the expected behavior. Should I create a regex that doesn't produce the additional blank strings?

Reply:
Yes, this is the expected behavior. Should I create a regex that doesn't produce the additional blank strings?

Reply:
I don't think the regex should be matching blank strings. It also looks like something you would have gotten from the internet :). I would suggest you try to use `re.compile` and see if you can match that URL.

Reply:
It was because of the presence of 4 capturing groups in the regex. I've converted the last 3 to non-capturing groups